homeassistant/components/tplink/entity.py

```diff
@@ -0,0 +1,70 @@
+"""Common code for tplink."""
+from __future__ import annotations
+
+from typing import Any, Callable, cast
+
+from kasa import SmartDevice
+
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.entity import DeviceInfo
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .coordinator import TPLinkDataUpdateCoordinator
+
+
+def async_refresh_after(func: Callable) -> Callable:
+    """Define a wrapper to refresh after."""
+
+    async def _async_wrap(
+        self: CoordinatedTPLinkEntity, *args: Any, **kwargs: Any
+    ) -> None:
+        await func(self, *args, **kwargs)
+        await self.coordinator.async_request_refresh_without_children()
+
+    return _async_wrap
+
+
+class CoordinatedTPLinkEntity(CoordinatorEntity):
+    """Common base class for all coordinated tplink entities."""
+
+    coordinator: TPLinkDataUpdateCoordinator
+
+    def __init__(
+        self, device: SmartDevice, coordinator: TPLinkDataUpdateCoordinator
+    ) -> None:
+        """Initialize the switch."""
+        super().__init__(coordinator)
+        self.device: SmartDevice = device
+
+    @property
+    def data(self) -> dict[str, Any]:
+        """Return data from DataUpdateCoordinator."""
+        data: dict[str, Any] = self.coordinator.data
+        return data
+
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return cast(str, self.device.device_id)
+
+    @property
+    def name(self) -> str:
+        """Return the name of the Smart Plug."""
+        return cast(str, self.device.alias)
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return information about the device."""
+        return {
+            "name": self.device.alias,
+            "model": self.device.model,
+            "manufacturer": "TP-Link",
+            # Note: mac instead of device_id here to connect subdevices to the main device
```

Comment:
We can always do both if we want.

Reply:
done in [`c7d9416` (#56632)](https://github.com/home-assistant/core/pull/56632/commits/c7d9416182690330e6992c1efa6c433572d787f3)